performance_manager/lib/rt_vehicle_positions.py

```diff
@@ -0,0 +1,277 @@
+
+    # store direction_id as Int64 [nullable]
+    df["direction_id"] = pandas.to_numeric(df["direction_id"]).astype("Int64")
+
+    # store start_time as seconds from start of day (Int64 [nullable])
+    df["start_time"] = (
+        df["start_time"].apply(start_time_to_seconds).astype("Int64")
+    )
+
+    # hash row data, exluding vehicle_timestamp column
+    # This hash is used to identify continuous series of records with the same
+    # categorical data
+    df["hash"] = df[list(set(df.columns) - {"vehicle_timestamp"})].apply(
+        lambda row: hash(tuple(row)), axis=1
+    )
+
+    # drop duplicates of hash and vehicle_timestamp columns (extraneous data)
+    df = df.drop_duplicates(subset=["vehicle_timestamp", "hash"])
+
+    return df
+
+
+def transform_vp_timestamps(df: pandas.DataFrame) -> pandas.DataFrame:
+    """
+    convert raw vehicle position data with one timestamp field into
+    vehicle position events with timestamp_start and timestamp_end
+    """
+    df = df.sort_values(by=["vehicle_id", "vehicle_timestamp"])
+
+    """
+    calculate mask to identify time series hash changes
+    this keeps first and last event for matching consecutive hash events
+    result is a 'start' event and 'end' event in consecutive rows for 
+    matching hash events
+    """
+    first_last_mask = (df["hash"] - df["hash"].shift(1) != 0) | (
+        df["hash"] - df["hash"].shift(-1) != 0
+    )
+    df = df.loc[first_last_mask, :]
+
+    """
+    transform vehicle_timestamp with matching consectuive hash events
+    into one record with timestamp_start and timestamp_end columns
+    if event is not part of consectuive matching events, then 
+    timestamp_start and timestamp_end will be same value
+    """
+    df.rename(columns={"vehicle_timestamp": "timestamp_start"}, inplace=True)
+    df["timestamp_end"] = numpy.where(
+        (df["hash"] - df["hash"].shift(-1) == 0),
+        df["timestamp_start"].shift(-1),
+        df["timestamp_start"],
+    ).astype("int64")
+
+    """
+    for matching consectuive hash events, drop 2nd event because timestamp
+    has been captured in timestamp_end value of previous event
+    """
+    drop_last_mask = df["hash"] - df["hash"].shift(1) != 0
+    df = df.loc[drop_last_mask, :].reset_index(drop=True)
+
+    df.insert(0, "id", numpy.nan)
+
+    # fill all dataframe na values with Python None (for DB writing)
+    df = df.fillna(numpy.nan).replace([numpy.nan], [None])
+
+    return df
+
+
+def get_utc_timestamp(path: str) -> float:
+    """
+    process datetime from s3 path return UTC timestamp
+    """
+    year = int(re.findall(r"year=(\d{4})", path)[0])
+    month = int(re.findall(r"month=(\d{1,2})", path)[0])
+    day = int(re.findall(r"day=(\d{1,2})", path)[0])
+    hour = int(re.findall(r"hour=(\d{1,2})", path)[0])
+    dt = datetime.datetime(
+        year=year, month=month, day=day, hour=hour, tzinfo=datetime.timezone.utc
+    )
+    return datetime.datetime.timestamp(dt)
```

Comment:
Get UTC timestamp from file path. This is used to identify the merge window for existing DB records. 5 minutes of either side of the hour from the file path. Might want to check our logic on bucketing/creating these paths to make sure it will align with this approach.